src/v/archival/archival_policy.cc

```diff
@@ -173,7 +173,8 @@ archival_policy::lookup_result archival_policy::find_segment(
 
     if (!closed) {
         auto kafka_start_offset = ot_state.from_log_offset(start_offset);
-        auto kafka_lso = ot_state.from_log_offset(adjusted_lso);
+        auto kafka_lso = ot_state.from_log_offset(
+          model::next_offset(adjusted_lso));
         if (kafka_start_offset >= kafka_lso) {
```

Comment:
nit: any reason we don't just adjust the condition here to be `>`? It seems more intuitive than having an "exclusive" lso. Or maybe we should rename `kafka_lso` to `kafka_first_unsable_offset` or somesuch.

Reply:
LSO is traditional kafka terminology so we use that. It is exclusive of the last message (e.g. in an empty partition it is 0). I agree that it is confusing and the fact that we have to adjust it to be inclusive and back only adds to the confusion.

But we can't just use inclusive offsets and substitute `>=` for `>` here to fix the bug (if I understood the suggestion correctly). E.g. if the last batch is [100-100] (1 msg long), then `start_offset` and `adjusted_lso` would be the same (both raw log offsets and kafka ones of course) so we can't distinguish between this batch being a data batch or an archival metadata batch. To do that we need to offset-translate 100 and 101 - if the resulting kafka offsets are the same this means that the batch is non-data and we shouldn't upload it.

Reply:
Thanks for explaining! So the idea of an inclusive LSO is just a Redpanda concept, even though Kafka's is exclusive?

I don't feel strongly about any of the suggestions, but I'm still confused, given these are just ints, it seems like there shouldn't be a difference between
`kafka start_offset >= translate(adjusted_lso) + 1` and `kafka_start_offset > translate(adjusted_lso)`
Or am I missing something?

Reply:
> there shouldn't be a difference between
`kafka start_offset >= translate(adjusted_lso) + 1 and kafka_start_offset > translate(adjusted_lso)`

Yeah, but we do `translate(adjusted_lso + 1)`, not `translate(adjusted_lso) + 1` which is quite different!

> So the idea of an inclusive LSO is just a Redpanda concept, even though Kafka's is exclusive?

The "adjusted lso" concept is local to this function (I guess for some reason it is just more convenient here or this is an artifact of code evolution). There are some comments [here](https://github.com/redpanda-data/redpanda/blob/dev/src/v/archival/archival_policy.cc#L505). Probably it would be even better to not talk about LSO here at all, just "inclusive end offset" :)

Reply:
Ah hah! That makes sense, thanks

>"inclusive end offset"
+1